common/src/main/java/com/tc/net/core/TCSocketEndpointReader.java

```diff
@@ -0,0 +1,251 @@
+  
+  public TCSocketEndpointReader() {
+    this.allocator = (s)->TCByteBufferFactory.getInstance(s);
+    this.returns = (b)->{};
+  }
+
+  public TCSocketEndpointReader(TCDirectByteBufferCache cache) {
+    this.allocator = (s)->cache.poll();
+    this.returns = cache::offer;
+  }
+  
+  // for testing
+  TCSocketEndpointReader(Function<Integer, TCByteBuffer> allocator, Consumer<TCByteBuffer> returns) {
+    this.allocator = allocator;
+    this.returns = returns;
+  }
+  
+  public TCReference readFromSocket(SocketEndpoint endpoint, int len) throws IOException {
+    LOGGER.debug("{} requesting:{} {} {}", endpoint, len, readBuffer, readTo);
+    if (readBuffer == null) {
+      TCByteBuffer newBuf = allocator.apply(len);
+      replaceCurrent(newBuf, createCompleteReference(newBuf), 0);
+    }
+    if (readBuffer.position() - readTo >= len) {
+    //  bytes are already off the network, return a slice for reading
+      current.stream().findFirst().get().position(readTo).limit(readTo + len);
+      TCReference ref = current.duplicate();
+      readTo += len;
+      LOGGER.debug("returning from cached bytes:{} {}",ref.available(),readBuffer);
+      return ref;
+    } else {
+      // need to fetch bytes from the network
+      LinkedList<TCByteBuffer> newBufs = new LinkedList<>();
+      int capacity = readBuffer.limit() - readTo;
+      // make sure there is enough capacity
+      while (capacity < len) {
+        TCByteBuffer next = allocator.apply(len - capacity);
+        newBufs.add(next);
+        capacity += next.limit();
+      }
+      // add the current buffer at the head
+      newBufs.addFirst(readBuffer);
+      int received = readBuffer.position() - readTo;
+      // read bytes from the network until the requested bytes are in
+      int rotations = 0;
+      while (received < len) {
+        try {
+          rotations += 1;
+          received += doRead(endpoint, newBufs);
+          LOGGER.debug("rotation:{} received:{}", rotations, received);
+        } catch (NoBytesAvailable no) {
+      //  no bytes in the channel, peer may not have written yet, 
+      //  if no bytes are cached, return null here and let the 
+      //  caller call again if needed
+          if (received == 0) {
+            newBufs.removeFirst();
+            if (!newBufs.isEmpty()) {
+              newBufs.forEach(b->returns.accept(b.reInit()));
+            }
+            LOGGER.debug("returning null");
+            return null;
+          } else {
+       // still expecting bytes, might be something wrong here.
+       // go around again after a short pause, maybe should consider
+       // failing at some point
+            if (rotations > 10) {
+              // give up, something is wrong
+              throw new IOException("incomplete bytes in channel");
+            } else {
+              try {
+                Thread.sleep(500);
+              } catch (InterruptedException ie) {
+                throw new IOException(ie);
```

Comment:
If this code can run in a client thread, it needs to re-assert the interrupt.

Reply:
done